### DISCUSSION TITLE
Practice Problem - Design chess

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#12] [General]
What?
Create the design for an online two-player chess.

Why?
Game problems.
Very popular for design and machine coding rounds.

How?
Follow the instructions [here](https://github.com/kanmaytacker/design-questions/blob/master/practice-problems/02-design-chess-worksheet.md).

**Comment by Faizahmadfaiz** [comment]
**Here is my first pass of the class diagram**._( I did it yesterday with all - requirement gathering part, use case daigram and class diagram but it got deleted by mistake. So now I only did class diagram)_

## Design Chess

### Class Diagram

```mermaid
classDiagram
class Game {
    Player[] players
    Board board
    Move[] moves
    +makeMove()
    +checkWinner()
}

class Player {
    -String colour
    +play(Piece, from: Cell, to: Cell)
}

class HumanPlayer {
    -String name
}

class BotPlayer {
    -int level
}

class Board {
    -int rows
    -int cols
    -Cell[][] cells
}

%% how to show empty cells?
class Cell {
    -String colour
    -Piece piece
    -int row
    -int col
}

%% How to show how different pieces move
%% steps and directions
class Piece {
    -String colour
}

class Move {
    -Player player
    -Piece piece
    -Cell from
    -Cell to
}

class King {
    
}

class Queen {
    
}

class Bishop {
    
}

class Rook {
    
}

class Knight {
    
}

class Pawn {
    
}

Game --o Player
Game --o Board

Player <-- HumanPlayer
Player <-- BotPlayer

Board --o Cell

Cell --o Piece


Move --o Player
Move --o Piece

Piece <-- King
Piece <-- Queen
Piece <-- Bishop
Piece <-- Rook
Piece <-- Knight
Piece <-- Pawn
```

I was not able to think -
* how to show Cell with no piece in class diagram.
* how to show behaviour of different pieces like the steps they can move and in which directions.
(Now I will see some videos and after that I will post here my second pass of the class diagram).

@kanmaytacker @all

**Reply by Faizahmadfaiz** [reply]
https://github.com/Faizahmadfaiz/chess/tree/main
Please find updated class diagram and partial code on this repo. Please give your feedbacks to improve. There are many missing things in code. Also I was not able come up again with these 2 things which I have pointed in above comment.

**Reply by kanmaytacker** [reply]
@Faizahmadfaiz First off, this is very close to the optimal design. Some good points of the design -
* Having a separate `Move` class. This can be used for storing a history later
* Having a generic class `Piece`. Is this an abstract class?

Missing things -
* You would need a status for each piece. For instance, if a piece is killed, we might want to show that.
* Similarly, how would represent that there is a check or checkmate happening. You might need a status again

Further optimisations - 
* Is it feasible to not have a separate class for each piece? This would depend on the requirements though.
* A generic way to have different behaviours for each piece. 

 Answering your questions below -
* `how to show Cell with no piece in class diagram` - Having a null value for piece is acceptable, however to optimise it further you could have a mapping table similar to [GroupExpense](https://github.com/kanmaytacker/design-questions/blob/splitwise-models/splitwise/src/main/java/com/scaler/splitwise/models/GroupExpense.java) in Splitwise.
*  `how to show behaviour of different pieces like the steps they can move and in which directions` - Each type of piece might have a separate behaviour. What would be great is to have a generic strategy that can be extended by different moving strategies.

**Comment by sastepankaj** [comment]
Hi,
Please find my class diagram for Chess. Thanks.

![Design Chess](https://user-images.githubusercontent.com/13418342/202793036-1393e5d9-df6d-4338-8319-3eda962b9c26.jpg)
